src/main/java/com/squareup/tools/maven/resolution/ProxyHelper.kt

```diff
@@ -0,0 +1,150 @@
+package com.squareup.tools.maven.resolution
+
+import okhttp3.Authenticator
+import okhttp3.Credentials
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.Route
+import java.net.URI
+import java.net.Proxy
+import java.net.InetSocketAddress
+import java.util.regex.Pattern
+
+sealed class ProxyEnvParseResult {
+  object NoConfig : ProxyEnvParseResult()
+
+  data class Error(val errorMessage: String) : ProxyEnvParseResult()
+
+  data class ProxyConfig(
+    val protocol: String,
+    val username: String?,
+    val password: String?,
+    val hostname: String,
+    val port: Int
+  ) : ProxyEnvParseResult() {
+    override fun toString(): String {
+      return "${protocol}://${hostname}:${port}"
+    }
+  }
+}
+
+sealed class ProxyExemptParseResult {
+  object NoProxyConfigured : ProxyExemptParseResult()
+
+  data class Error(val errorMessage: String) : ProxyExemptParseResult()
+  data class Exempt(val url: String) : ProxyExemptParseResult()
+  data class NotExempt(val proxy: Proxy) : ProxyExemptParseResult()
+}
+
+object ProxyUtils {
+
+  fun getConfig(proxyAddress: String = System.getenv("https_proxy") ?: System.getenv("HTTPS_PROXY")): ProxyEnvParseResult {
+    val urlPattern =
+      Pattern.compile("^(https?)://(([^:@]+?)(?::([^@]+?))?@)?([^:]+)(?::(\\d+))?/?$")
+    val matcher = urlPattern.matcher(proxyAddress)
```

Comment:
Weird. How did this compile? getenv() returns a `String?` and matcher takes `String`.

Reply:
I'm on Kotlin 1.3.50...something must have changed since then! :)